### ISSUE TITLE
`mul_add` suggestion from `suboptimal_flops` applies unary minus to unsigned integer, leading to compiler error

### ISSUE THREAD

**Issue by ijijn** [rust-lang/rust-clippy#11559]
### Summary

This issue is triggered when subtracting a cast from an unsigned type in a `mul_add` context. Once again, parentheses around this cast to the rescue 😅 

### Reproducer

I tried this code:

```rust
#[must_use]
pub fn proportion(start: f64, drop: f64, index: usize) -> f64 {
    #[allow(clippy::cast_precision_loss)]
    let result = start - (drop * index as f64);

    result
}
```

I expected to see this happen:

```rust
#[must_use]
pub fn proportion(start: f64, drop: f64, index: usize) -> f64 {
    #[allow(clippy::cast_precision_loss)]
    let result = drop.mul_add(-(index as f64), start);

    result
}
```

Instead, this happened:

```rust
#[must_use]
pub fn proportion(start: f64, drop: f64, index: usize) -> f64 {
    #[allow(clippy::cast_precision_loss)]
    let result = drop.mul_add(-index as f64, start);

    result
}
```

which is a compiler error.

**Comment by asukaminato0721** [comment]
+1, my case is

```rs
error: multiply and add expressions can be calculated more efficiently and accurately
   --> src/optimal_retention.rs:310:25
    |
310 |             *new_diff = (w[4] - w[5] * (rating as f64 - 3.0)).clamp(1.0, 10.0);
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using: `w[5].mul_add(-rating as f64 - 3.0, w[4])`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suboptimal_flops
```

but it should be

```rs
*new_diff = (w[5].mul_add(-(rating as f64 - 3.0), w[4])).clamp(1.0, 10.0);
```
